mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp

```diff
@@ -79,22 +80,36 @@ createVariablesForResults(T op, const TypeConverter *typeConverter,
 
 // Create a series of assign ops assigning given values to given variables at
 // the current insertion point of given rewriter.
-static void assignValues(ValueRange values, SmallVector<Value> &variables,
-                         ConversionPatternRewriter &rewriter, Location loc) {
+static void assignValues(ValueRange values, ValueRange variables,
+                         ConversionPatternRewriter &rewriter, Location loc,
+                         const TypeConverter *typeConverter = nullptr) {
   for (auto [value, var] : llvm::zip(values, variables))
     rewriter.create<emitc::AssignOp>(loc, var, value);
 }
 
-static void lowerYield(SmallVector<Value> &resultVariables,
-                       ConversionPatternRewriter &rewriter,
-                       scf::YieldOp yield) {
+static void lowerYield(ValueRange resultVariables,
+                       ConversionPatternRewriter &rewriter, scf::YieldOp yield,
+                       const TypeConverter *typeConverter) {
   Location loc = yield.getLoc();
-  ValueRange operands = yield.getOperands();
 
   OpBuilder::InsertionGuard guard(rewriter);
   rewriter.setInsertionPoint(yield);
 
-  assignValues(operands, resultVariables, rewriter, loc);
+  SmallVector<Value> yieldOperands;
+  for (auto originalOperand : yield.getOperands()) {
+    Value operand = originalOperand;
+
+    if (typeConverter && !typeConverter->isLegal(operand.getType())) {
+      Type resultType = typeConverter->convertType(operand.getType());
+      auto castToTarget =
```

Comment:
Instead of explicitly creating the unrealized cast, would `ConversionPatternRewriter::getRemappedValue(operand)` maybe allow to get the new operand? (Either as created before or via the typeconverter materialization)

Reply:
Ah! That seems to also work, be simpler and more converter friendly. will update.

Reply:
Fixed